server/src/Persistence.hs

```diff
@@ -0,0 +1,175 @@
+type RawJournalData = [SBS.ByteString]
+
+-- | Open or create the journal file
+openJournal :: FilePath -> ExceptT String IO Handle
+openJournal journalFile = ExceptT $ do
+  eitherHandle <- try $ do
+    h <- openBinaryFile journalFile ReadWriteMode
+    hSetBuffering h LineBuffering
+    pure h
+  case eitherHandle :: Either SomeException Handle of
+    Left exc -> pure $ Left $ "Failed to open journal file: " ++ show exc
+    Right fileHandle -> pure $ Right fileHandle
+
+-- | Read the modifications from the journal file, apply them and sync again.
+-- This should be done when loading the database from disk.
+recoverJournal :: PersistentValue -> ExceptT String IO ()
+recoverJournal pval = for_ (pvJournal pval) $ \journalHandle -> ExceptT $ fmap formatErr $ try $ do
+  -- read modifications from the beginning
+  journalLines <- do
+    hSeek journalHandle AbsoluteSeek 0
+    readJournal journalHandle
+  -- parse and apply modifications from journal
+  let (errs, ops) = parseJournalData journalLines
+
+  when (not $ null ops) $ do
+    pcLog (pvConfig pval) "Journal not empty, recovering"
+
+  when (not $ null errs) $ do
+    let msg = Text.intercalate "\n" $ "Failed to recover some journal entries:" : map Text.pack errs
+    pcLog (pvConfig pval) msg
+
+  atomically $ do
+    modifyTVar' (pvValue pval) (replayModifications ops)
+    writeTVar (pvIsDirty pval) True
+  -- syncing takes care of cleaning the journal
+  sync pval
+
+  when (not $ null ops) $ do
+    pcLog (pvConfig pval) "Journal replayed"
+
+  where
+    formatErr :: Either SomeException a -> Either String a
+    formatErr (Left exc) = Left $ "Failed to read journal: " ++ show exc
+    formatErr (Right x)  = Right x
+
+-- | Parse journal data in a list of modifications and a list of errors for entries that could not be parsed.
+parseJournalData :: RawJournalData -> ([String], [Store.Modification])
+parseJournalData = partitionEithers . map Aeson.eitherDecodeStrict
+
+-- | Replay a list of modifications from an initial value.
+replayModifications :: [Store.Modification] -> Store.Value -> Store.Value
+replayModifications ops initial = foldl' (flip Store.applyModification) initial ops
+
+-- | Read and decode the data file
+readData :: FilePath -> ExceptT String IO Store.Value
+readData filePath = ExceptT $ do
+  eitherEncodedValue <- try $ withFile filePath ReadMode SBS.hGetContents
+  case (eitherEncodedValue :: Either SomeException SBS.ByteString) of
+    Left exc -> pure $ Left $ "Failed to read the data from disk: " ++ show exc
+    Right encodedValue -> do
+      case Aeson.eitherDecodeStrict encodedValue of
+        Left msg  -> pure $ Left $ "Failed to decode the initial data: " ++ show msg
+        Right value -> pure $ Right $ value
+
+-- | Read the journal file.
+readJournal :: Handle -> IO RawJournalData
+readJournal h = go where
+  go = do
+    eof <- hIsEOF h
+    if eof
+      then pure []
+      else (:) <$> SBS8.hGetLine h <*> go
```

Comment:
Did you test this function with a large journal? It seems like it is not tail-recursive.

Reply:
The problem with `hGetContents` is, that it closes the file handle when it's done. But I would like to keep the file handle to the journal open all the time, because it makes stuff simpler if one only has to remember to open the file once.

I think I could add an `unsafeInterleaveIO` though (what `hGetContents` uses internally as well) in order to do lazy IO. Also, that would reduce the memory load when recovering a very large journal. The more elegant way would of course to use one of the numerous streaming libraries for Haskell, but that's overkill for one-time-use I'd say.

Reply:
Fixed. Didn't get it to work with `unsafeInterlaveIO`, but now I have a tail-recursive function folding over all the lines and replaying the journal on the go. Now it can replay the journal in constant memory as well.